### PR TITLE
First pass at asyncifying

### PR DESCRIPTION
**Pull request by rylev**
Not only does this make most things async, but it also now uses the latest version of `kube`

### CODE REVIEW

**Comment by radu-matei** [comment]
I can compile and all tests pass locally on macOS - but CI seems to fail:

```
2020-03-05T17:43:25.0299649Z    Compiling krustlet v0.1.0 (/home/runner/work/krustlet/krustlet)
2020-03-05T17:43:25.9113785Z error[E0195]: lifetime parameters or bounds on method `add` do not match the trait declaration
2020-03-05T17:43:25.9115717Z    --> crates/kubelet/src/kubelet.rs:474:15
2020-03-05T17:43:25.9115915Z     |
2020-03-05T17:43:25.9116042Z 149 | #[async_trait]
2020-03-05T17:43:25.9116372Z     | -------------- lifetimes in impl do not match this method in trait
2020-03-05T17:43:25.9116514Z ...
2020-03-05T17:43:25.9117261Z 474 |         fn add(&self, _pod: Pod, _client: APIClient) -> Result<(), failure::Error> {
2020-03-05T17:43:25.9117448Z     |               ^ lifetimes do not match method in trait
2020-03-05T17:43:25.9117522Z 
2020-03-05T17:43:25.9117665Z error[E0195]: lifetime parameters or bounds on method `modify` do not match the trait declaration
2020-03-05T17:43:25.9117968Z    --> crates/kubelet/src/kubelet.rs:477:18
2020-03-05T17:43:25.9118101Z     |
2020-03-05T17:43:25.9118227Z 149 | #[async_trait]
2020-03-05T17:43:25.9118535Z     | -------------- lifetimes in impl do not match this method in trait
2020-03-05T17:43:25.9118687Z ...
2020-03-05T17:43:25.9119229Z 477 |         fn modify(&self, _pod: Pod, _client: APIClient) -> Result<(), failure::Error> {
2020-03-05T17:43:25.9119391Z     |                  ^ lifetimes do not match method in trait
2020-03-05T17:43:25.9119484Z 
2020-03-05T17:43:25.9119622Z error[E0195]: lifetime parameters or bounds on method `status` do not match the trait declaration
2020-03-05T17:43:25.9119923Z    --> crates/kubelet/src/kubelet.rs:480:18
2020-03-05T17:43:25.9120054Z     |
2020-03-05T17:43:25.9120155Z 149 | #[async_trait]
2020-03-05T17:43:25.9120617Z     | -------------- lifetimes in impl do not match this method in trait
2020-03-05T17:43:25.9120764Z ...
2020-03-05T17:43:25.9121342Z 480 |         fn status(&self, _pod: Pod, _client: APIClient) -> Result<Status, failure::Error> {
2020-03-05T17:43:25.9121505Z     |                  ^ lifetimes do not match method in trait
2020-03-05T17:43:25.9121599Z 
2020-03-05T17:43:26.5510552Z error[E0308]: mismatched types
2020-03-05T17:43:26.5511854Z    --> crates/kubelet/src/kubelet.rs:452:21
2020-03-05T17:43:26.5512217Z     |
2020-03-05T17:43:26.5512682Z 452 |             client: reqwest::Client::new(),
2020-03-05T17:43:26.5513069Z     |                     ^^^^^^^^^^^^^^^^^^^^^^ expected struct `reqwest::async_impl::client::Client`, found struct `reqwest::client::Client`
2020-03-05T17:43:26.5513366Z 
2020-03-05T17:43:26.7494950Z error[E0308]: mismatched types
2020-03-05T17:43:26.7496019Z    --> crates/kubelet/src/kubelet.rs:574:17
2020-03-05T17:43:26.7496341Z     |
2020-03-05T17:43:26.7496647Z 574 |                 labels,
2020-03-05T17:43:26.7496907Z     |                 ^^^^^^
2020-03-05T17:43:26.7497351Z     |                 |
2020-03-05T17:43:26.7497816Z     |                 expected enum `std::option::Option`, found struct `std::collections::BTreeMap`
2020-03-05T17:43:26.7498165Z     |                 help: try using a variant of the expected enum: `Some(labels)`
2020-03-05T17:43:26.7498475Z     |
2020-03-05T17:43:26.7498820Z     = note: expected enum `std::option::Option<std::collections::BTreeMap<_, _>>`
2020-03-05T17:43:26.7499354Z              found struct `std::collections::BTreeMap<_, _>`
2020-03-05T17:43:26.7499625Z 
2020-03-05T17:43:26.8282774Z error[E0422]: cannot find struct, variant or union type `KubePod` in this scope
2020-03-05T17:43:26.8283888Z    --> crates/wascc-provider/src/lib.rs:283:24
2020-03-05T17:43:26.8284212Z     |
2020-03-05T17:43:26.8284493Z 283 |         let mut mock = KubePod {
2020-03-05T17:43:26.8284816Z     |                        ^^^^^^^ not found in this scope
2020-03-05T17:43:26.8285040Z 
2020-03-05T17:43:26.8322315Z error[E0422]: cannot find struct, variant or union type `PodSpec` in this scope
2020-03-05T17:43:26.8323100Z    --> crates/wascc-provider/src/lib.rs:296:21
2020-03-05T17:43:26.8323437Z     |
2020-03-05T17:43:26.8323735Z 296 |         mock.spec = PodSpec {
2020-03-05T17:43:26.8324044Z     |                     ^^^^^^^ not found in this scope
2020-03-05T17:43:26.8324345Z     |
2020-03-05T17:43:26.8324675Z help: possible candidate is found in another module, you can import it into scope
2020-03-05T17:43:26.8324986Z     |
2020-03-05T17:43:26.8325420Z 253 |     use k8s_openapi::api::core::v1::PodSpec;
2020-03-05T17:43:26.8325705Z     |
2020-03-05T17:43:26.8325949Z 
2020-03-05T17:43:26.8361740Z error[E0422]: cannot find struct, variant or union type `PodSpec` in this scope
2020-03-05T17:43:26.8362396Z    --> crates/wascc-provider/src/lib.rs:302:21
2020-03-05T17:43:26.8362702Z     |
2020-03-05T17:43:26.8362970Z 302 |         mock.spec = PodSpec {
2020-03-05T17:43:26.8363270Z     |                     ^^^^^^^ not found in this scope
2020-03-05T17:43:26.8363553Z     |
2020-03-05T17:43:26.8363837Z help: possible candidate is found in another module, you can import it into scope
2020-03-05T17:43:26.8364127Z     |
2020-03-05T17:43:26.8364472Z 253 |     use k8s_openapi::api::core::v1::PodSpec;
2020-03-05T17:43:26.8364737Z     |
2020-03-05T17:43:26.8364963Z 
2020-03-05T17:43:26.8413758Z warning: unused import: `kubelet::pod::Pod`
2020-03-05T17:43:26.8414641Z    --> crates/wascc-provider/src/lib.rs:254:9
2020-03-05T17:43:26.8414954Z     |
2020-03-05T17:43:26.8415270Z 254 |     use kubelet::pod::Pod;
2020-03-05T17:43:26.8415554Z     |         ^^^^^^^^^^^^^^^^^
2020-03-05T17:43:26.8415829Z     |
2020-03-05T17:43:26.8416104Z     = note: `#[warn(unused_imports)]` on by default
2020-03-05T17:43:26.8416322Z 
2020-03-05T17:43:26.9360973Z error[E0308]: mismatched types
2020-03-05T17:43:26.9362106Z    --> crates/kubelet/src/kubelet.rs:575:17
2020-03-05T17:43:26.9362597Z     |
2020-03-05T17:43:26.9362905Z 575 |                 annotations,
2020-03-05T17:43:26.9363187Z     |                 ^^^^^^^^^^^
2020-03-05T17:43:26.9363461Z     |                 |
2020-03-05T17:43:26.9363907Z     |                 expected enum `std::option::Option`, found struct `std::collections::BTreeMap`
2020-03-05T17:43:26.9364232Z     |                 help: try using a variant of the expected enum: `Some(annotations)`
2020-03-05T17:43:26.9364533Z     |
2020-03-05T17:43:26.9364855Z     = note: expected enum `std::option::Option<std::collections::BTreeMap<_, _>>`
2020-03-05T17:43:26.9365200Z              found struct `std::collections::BTreeMap<_, _>`
2020-03-05T17:43:26.9365428Z 
2020-03-05T17:43:27.1235204Z error[E0308]: mismatched types
2020-03-05T17:43:27.1236498Z    --> crates/kubelet/src/kubelet.rs:576:17
2020-03-05T17:43:27.1236847Z     |
2020-03-05T17:43:27.1237168Z 576 |                 name,
2020-03-05T17:43:27.1237466Z     |                 ^^^^
2020-03-05T17:43:27.1237789Z     |                 |
2020-03-05T17:43:27.1238260Z     |                 expected enum `std::option::Option`, found struct `std::string::String`
2020-03-05T17:43:27.1238604Z     |                 help: try using a variant of the expected enum: `Some(name)`
2020-03-05T17:43:27.1238915Z     |
2020-03-05T17:43:27.1239256Z     = note: expected enum `std::option::Option<std::string::String>`
2020-03-05T17:43:27.1239616Z              found struct `std::string::String`
2020-03-05T17:43:27.1240092Z 
2020-03-05T17:43:27.3031269Z error[E0308]: mismatched types
2020-03-05T17:43:27.3032379Z    --> crates/kubelet/src/kubelet.rs:573:23
2020-03-05T17:43:27.3032707Z     |
2020-03-05T17:43:27.3032995Z 573 |               metadata: ObjectMeta {
2020-03-05T17:43:27.3033276Z     |  _______________________^
2020-03-05T17:43:27.3033551Z 574 | |                 labels,
2020-03-05T17:43:27.3033826Z 575 | |                 annotations,
2020-03-05T17:43:27.3034102Z 576 | |                 name,
2020-03-05T17:43:27.3034526Z 577 | |                 namespace,
2020-03-05T17:43:27.3034962Z 578 | |                 ..Default::default()
2020-03-05T17:43:27.3035251Z 579 | |             },
2020-03-05T17:43:27.3035602Z     | |_____________^ expected enum `std::option::Option`, found struct `k8s_openapi::v1_17::apimachinery::pkg::apis::meta::v1::object_meta::ObjectMeta`
2020-03-05T17:43:27.3035918Z     |
2020-03-05T17:43:27.3036258Z     = note: expected enum `std::option::Option<k8s_openapi::v1_17::apimachinery::pkg::apis::meta::v1::object_meta::ObjectMeta>`
2020-03-05T17:43:27.3036627Z              found struct `k8s_openapi::v1_17::apimachinery::pkg::apis::meta::v1::object_meta::ObjectMeta`
2020-03-05T17:43:27.3036935Z help: try using a variant of the expected enum
2020-03-05T17:43:27.3037209Z     |
2020-03-05T17:43:27.3037485Z 573 |             metadata: Some(ObjectMeta {
2020-03-05T17:43:27.3037762Z 574 |                 labels,
2020-03-05T17:43:27.3038038Z 575 |                 annotations,
2020-03-05T17:43:27.3038312Z 576 |                 name,
2020-03-05T17:43:27.3038591Z 577 |                 namespace,
2020-03-05T17:43:27.3038900Z 578 |                 ..Default::default()
2020-03-05T17:43:27.3039160Z   ...
2020-03-05T17:43:27.3039389Z 
2020-03-05T17:43:27.3702002Z error[E0599]: no method named `expect` found for type `std::pin::Pin<std::boxed::Box<dyn std::future::Future<Output = std::result::Result<(), failure::Error>> + std::marker::Send>>` in the current scope
2020-03-05T17:43:27.3752882Z    --> crates/wascc-provider/src/lib.rs:259:25
2020-03-05T17:43:27.3753280Z     |
2020-03-05T17:43:27.3753535Z 259 |         provider.init().expect("HTTP capability is registered");
2020-03-05T17:43:27.3753849Z     |                         ^^^^^^ method not found in `std::pin::Pin<std::boxed::Box<dyn std::future::Future<Output = std::result::Result<(), failure::Error>> + std::marker::Send>>`
2020-03-05T17:43:27.3754199Z 
2020-03-05T17:43:27.3923886Z error: aborting due to 4 previous errors
2020-03-05T17:43:27.3924306Z 
2020-03-05T17:43:27.3928662Z Some errors have detailed explanations: E0422, E0599.
2020-03-05T17:43:27.3934001Z For more information about an error, try `rustc --explain E0422`.
2020-03-05T17:43:27.3999975Z error: could not compile `wascc-provider`.
2020-03-05T17:43:27.4001399Z warning: build failed, waiting for other jobs to finish...
2020-03-05T17:43:27.4768368Z error[E0308]: mismatched types
2020-03-05T17:43:27.4769799Z    --> crates/kubelet/src/kubelet.rs:580:19
2020-03-05T17:43:27.4770421Z     |
2020-03-05T17:43:27.4770975Z 580 |               spec: PodSpec {
2020-03-05T17:43:27.4771497Z     |  ___________________^
2020-03-05T17:43:27.4772045Z 581 | |                 service_account_name: Some("svc".to_string()),
2020-03-05T17:43:27.4772618Z 582 | |                 ..Default::default()
2020-03-05T17:43:27.4773115Z 583 | |             },
2020-03-05T17:43:27.4773708Z     | |_____________^ expected enum `std::option::Option`, found struct `k8s_openapi::v1_17::api::core::v1::pod_spec::PodSpec`
2020-03-05T17:43:27.4774271Z     |
2020-03-05T17:43:27.4774848Z     = note: expected enum `std::option::Option<k8s_openapi::v1_17::api::core::v1::pod_spec::PodSpec>`
2020-03-05T17:43:27.4775450Z              found struct `k8s_openapi::v1_17::api::core::v1::pod_spec::PodSpec`
2020-03-05T17:43:27.4776009Z help: try using a variant of the expected enum
2020-03-05T17:43:27.4776526Z     |
2020-03-05T17:43:27.4777033Z 580 |             spec: Some(PodSpec {
2020-03-05T17:43:27.4777582Z 581 |                 service_account_name: Some("svc".to_string()),
2020-03-05T17:43:27.4778415Z 582 |                 ..Default::default()
2020-03-05T17:43:27.4778941Z 583 |             }),
2020-03-05T17:43:27.4779447Z     |
2020-03-05T17:43:27.4779864Z 
2020-03-05T17:43:27.4852047Z error[E0599]: no method named `get` found for type `std::pin::Pin<std::boxed::Box<dyn failure::core::future::future::Future<Output = std::collections::HashMap<std::string::String, std::string::String>> + std::marker::Send>>` in the current scope
2020-03-05T17:43:27.4852828Z    --> crates/kubelet/src/kubelet.rs:595:17
2020-03-05T17:43:27.4853160Z     |
2020-03-05T17:43:27.4853457Z 595 |             env.get("first").expect("key first should exist").as_str()
2020-03-05T17:43:27.4853765Z     |                 ^^^ help: there is a method with a similar name: `set`
2020-03-05T17:43:27.4853992Z 
2020-03-05T17:43:27.4927097Z error[E0599]: no method named `get` found for type `std::pin::Pin<std::boxed::Box<dyn failure::core::future::future::Future<Output = std::collections::HashMap<std::string::String, std::string::String>> + std::marker::Send>>` in the current scope
2020-03-05T17:43:27.4927906Z    --> crates/kubelet/src/kubelet.rs:600:17
2020-03-05T17:43:27.4928233Z     |
2020-03-05T17:43:27.4928519Z 600 |             env.get("second").expect("metadata.labels.label").as_str()
2020-03-05T17:43:27.4928820Z     |                 ^^^ help: there is a method with a similar name: `set`
2020-03-05T17:43:27.4929061Z 
2020-03-05T17:43:27.4996189Z error[E0599]: no method named `get` found for type `std::pin::Pin<std::boxed::Box<dyn failure::core::future::future::Future<Output = std::collections::HashMap<std::string::String, std::string::String>> + std::marker::Send>>` in the current scope
2020-03-05T17:43:27.4996950Z    --> crates/kubelet/src/kubelet.rs:604:17
2020-03-05T17:43:27.4997269Z     |
2020-03-05T17:43:27.4997547Z 604 |             env.get("third")
2020-03-05T17:43:27.4997839Z     |                 ^^^ help: there is a method with a similar name: `set`
2020-03-05T17:43:27.4998098Z 
2020-03-05T17:43:27.5066862Z error[E0599]: no method named `get` found for type `std::pin::Pin<std::boxed::Box<dyn failure::core::future::future::Future<Output = std::collections::HashMap<std::string::String, std::string::String>> + std::marker::Send>>` in the current scope
2020-03-05T17:43:27.5067714Z    --> crates/kubelet/src/kubelet.rs:608:35
2020-03-05T17:43:27.5068060Z     |
2020-03-05T17:43:27.5068594Z 608 |         assert_eq!("my-name", env.get("NAME").expect("metadata.name").as_str());
2020-03-05T17:43:27.5069674Z     |                                   ^^^ help: there is a method with a similar name: `set`
2020-03-05T17:43:27.5069946Z 
2020-03-05T17:43:27.5139485Z error[E0599]: no method named `get` found for type `std::pin::Pin<std::boxed::Box<dyn failure::core::future::future::Future<Output = std::collections::HashMap<std::string::String, std::string::String>> + std::marker::Send>>` in the current scope
2020-03-05T17:43:27.5140274Z    --> crates/kubelet/src/kubelet.rs:611:17
2020-03-05T17:43:27.5140540Z     |
2020-03-05T17:43:27.5140773Z 611 |             env.get("NAMESPACE").expect("metadata.namespace").as_str()
2020-03-05T17:43:27.5141017Z     |                 ^^^ help: there is a method with a similar name: `set`
2020-03-05T17:43:27.5141184Z 
2020-03-05T17:43:27.5210787Z error[E0599]: no method named `get` found for type `std::pin::Pin<std::boxed::Box<dyn failure::core::future::future::Future<Output = std::collections::HashMap<std::string::String, std::string::String>> + std::marker::Send>>` in the current scope
2020-03-05T17:43:27.5211561Z    --> crates/kubelet/src/kubelet.rs:613:38
2020-03-05T17:43:27.5211839Z     |
2020-03-05T17:43:27.5212096Z 613 |         assert_eq!("10.21.77.2", env.get("POD_IP").expect("pod_ip").as_str());
2020-03-05T17:43:27.5212368Z     |                                      ^^^ help: there is a method with a similar name: `set`
2020-03-05T17:43:27.5212575Z 
2020-03-05T17:43:27.5279432Z error[E0599]: no method named `get` found for type `std::pin::Pin<std::boxed::Box<dyn failure::core::future::future::Future<Output = std::collections::HashMap<std::string::String, std::string::String>> + std::marker::Send>>` in the current scope
2020-03-05T17:43:27.5280263Z    --> crates/kubelet/src/kubelet.rs:614:38
2020-03-05T17:43:27.5280518Z     |
2020-03-05T17:43:27.5280756Z 614 |         assert_eq!("10.21.77.1", env.get("HOST_IP").expect("host_ip").as_str());
2020-03-05T17:43:27.5281014Z     |                                      ^^^ help: there is a method with a similar name: `set`
2020-03-05T17:43:27.5281187Z 
2020-03-05T17:43:27.5581939Z error: aborting due to 16 previous errors

```

File: crates/kubelet/src/kubelet.rs [hunk 1]
```diff
@@ -79,42 +84,41 @@ impl<T: 'static + Provider + Sync + Send + Clone> Kubelet<T> {
...
-            let informer = Informer::new(pod_client)
-                .init()
-                .expect("informer init failed");
+            let informer = Informer::new(client, ListParams::default(), Resource::all::<Pod>());
             loop {
-                informer.poll().expect("informer poll failed");
-                while let Some(event) = informer.pop() {
+                let mut stream = informer.poll().await.expect("informer poll failed").boxed();
+                while let Some(event) = stream.try_next().await.unwrap() {
                     // TODO: We need to spawn threads (or do something similar)
```
Comment (thomastaylor312) [inline crates/kubelet/src/kubelet.rs]: I think this comment can go now

File: crates/kubelet/src/kubelet.rs [hunk 2]
```diff
@@ -123,12 +127,12 @@ impl<T: 'static + Provider + Sync + Send + Clone> Kubelet<T> {
         });
 
         // Start the webserver
-        start_webserver(self.provider.clone(), &address)?;
+        start_webserver(self.provider.clone(), &address).await?;
 
         // Join the threads
```
Comment (thomastaylor312) [inline crates/kubelet/src/kubelet.rs]: Nit: no longer need this comment I think

File: crates/kubelet/src/kubelet.rs [hunk 3]
```diff
@@ -265,102 +279,122 @@ pub trait Provider {
...
+    ns: &str,
+    fields: &HashMap<String, String>,
+) -> String {
+    if let Some(env_src) = env_var.value_from.as_ref() {
+        // ConfigMaps
+        if let Some(cfkey) = env_src.config_map_key_ref.as_ref() {
+            let name = cfkey.name.as_deref().unwrap_or("");
+            match Api::<ConfigMap>::namespaced(client, ns).get(name).await {
+                Ok(cfgmap) => {
+                    // I am not totally clear on what the outcome should
```
Comment (thomastaylor312) [inline crates/kubelet/src/kubelet.rs]: Note for the future: I think you are supposed to error if this happens if I remember correctly

File: crates/kubelet/src/node.rs [hunk 4]
```diff
@@ -21,28 +21,30 @@ const NODE_NAME: &str = "krustlet";
 ///
 /// A node comes with a lease, and we maintain the lease to tell Kubernetes that the
 /// node remains alive and functional. Note that this will not work in
-/// versions of Kubernetes prior to 1.14.
-pub fn create_node(client: &APIClient, arch: &str) {
-    let node_client = Api::v1Node(client.clone());
-    let pp = PostParams::default();
+/// versions of Kubernetes prior to 1.  14.
```
Comment (thomastaylor312) [inline crates/kubelet/src/node.rs]: Nit: looks like there was an extra space added after `1.`

**Comment by radu-matei** [comment]
ref https://github.com/Arnavion/k8s-openapi/commit/35cbddfeb26726bc1a0c882a40584c262fb53726

**Pull request merged**